bundler/lib/bundler/checksum.rb

```diff
@@ -179,16 +185,14 @@ def inspect
       # that contain the same gem with different checksums.
       def replace(spec, checksum)
         return unless checksum
-
         lock_name = spec.name_tuple.lock_name
-        checksums = (store[lock_name] ||= {})
-        existing = checksums[checksum.algo]
-
-        # we assume only one source because this is used while building the index
-        if !existing || existing.sources.first == checksum.sources.first
-          checksums[checksum.algo] = checksum
-        else
-          register_checksum(lock_name, checksum)
+        @store_mutex.synchronize do
+          existing = fetch_checksum(lock_name, checksum.algo)
+          if !existing || existing.same_source?(checksum)
```

Comment:
The implementation of this method sounds slightly different than before? Also should the comment be carried over?

Reply:
I short-circuited. `register_checksum` was re-fetching the existing checksum when we already have it here.

Reply:
The comment is no longer correct. I made it so same_source checks for any matching source on existing, and the incoming one should always have only 1. I think the method name explains it well enough now.

Reply:
I meant the one you extracted, `same_source?`.

Reply:
I get it now!